verifier/src/commitment.rs

```diff
@@ -0,0 +1,173 @@
+use super::{error::Error, utils::compute_label_commitment, HashCommitment, LabelSeed};
+use serde::Serialize;
+use std::any::Any;
+
+/// A User's commitment to a portion of the notarized data
+#[derive(Serialize)]
+pub struct Commitment {
+    id: usize,
+    typ: CommitmentType,
+    direction: Direction,
+    // The index of this commitment in the Merkle tree of commitments
+    merkle_tree_index: usize,
+    // The actual commitment
+    commitment: HashCommitment,
+    // The absolute byte ranges within the notarized data. The committed data
+    // is located in those ranges.
+    ranges: Vec<Range>,
+}
+
+impl Commitment {
+    pub fn new(
+        id: usize,
+        typ: CommitmentType,
+        direction: Direction,
+        commitment: HashCommitment,
+        ranges: Vec<Range>,
+        merkle_tree_index: usize,
+    ) -> Self {
+        Self {
+            id,
+            typ,
+            direction,
+            commitment,
+            ranges,
+            merkle_tree_index,
+        }
+    }
+
+    /// Verifies this commitment against the opening. `extra_data` holds extra data specific
+    /// to the commitment type.
+    pub fn verify(
+        &self,
+        opening: &CommitmentOpening,
+        extra_data: Box<dyn Any>,
```

Comment:
> will just be included in the commitment variant

you meant `CommitmentOpening` variant, i assume

like this?

```rust
enum CommitmentOpening {
    /// Opening data for commitment of type [CommitmentType::labels_blake3]
    LabelsBlake3 {
        id: usize,
        opening: Vec<u8>,
        salt: Vec<u8>,
        extra_data: LabelSeed,
    },
}
```

Reply:
agreed, except I chose not to use the `Unchecked` approach on each type individually, so that all validations could be performed in one place of the code.
Otherwise we would have some validations on an individual type and other kinds of validations on the whole document.

Reply:
The suggested approach didn't work out because when the http parser wants to access the opening, currently it does:
`verified_doc.commitment_openings()[0].opening()`
but with the suggested approach the parser would have to know the type of the opening in order to extract the enum's variant and then call `.opening()` on the variant.

Reply:
It seems to make the most sense that the output of verification should be a struct containing only the useful data that will be used by the parser, i.e.

```rust
struct VerifiedTranscript {
    date: u64,
    dns_name: String,
    data: Vec<TranscriptSlice>
}
```

If you agree, i'll incorporate this into the Verifier.